File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> hash = new HashMap<>();
            
        int len = nums.length;
        
        for(int i = 0; i < len; i++) {
            if(!hash.containsKey(nums[i]))
                hash.put(nums[i], 1);
            else
                hash.put(nums[i], 1 + hash.get(nums[i]));
        }
        
        PriorityQueue<Map.Entry<Integer, Integer>> maxPQ =  
                             new PriorityQueue<>((a, b) -> b.getValue() - a.getValue()); 
                  
        
        for(Map.Entry<Integer, Integer> entry : hash.entrySet()) 
            maxPQ.add(entry);
        
        
        int[] ans = new int[k];
        
        for (int i = 0; i < k; i++) 
            ans[i] = maxPQ.poll().getKey();

        
        return ans;
        

    }
}

File: 0347-top-k-frequent-elements/0347-top-k-frequent-elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        counter = {}
        max_freq, ans, count = -1, [], 0 

        for e in nums:
            counter[e] = counter.get(e, 0) + 1
            max_freq = max(max_freq, counter[e])

        freq =[[] for _ in range(max_freq + 1)]
        
        for key, val in counter.items(): freq[val].append(key)
        
        for i in range(max_freq, 0, -1):
            for val in freq[i]:
                count += 1
                ans.append(val)
                if count == k: return ans

